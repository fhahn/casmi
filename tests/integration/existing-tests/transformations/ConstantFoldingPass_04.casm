// transform, cmdline "--cf"

CASM ConstantFoldingPass

init foo

function (static) bar : -> Int initially { 1234 }

function (static) baz : Int -> Int initially { 0 -> 3210, 1 -> 4321 }

function (static) qux : Int * Int -> Int initially { [5, 12] -> 1987 }

function (static) prog_mem : Int -> RuleRef initially
{
	1 -> @foo,
	2 -> @bbb,
	3 -> @ccc,
}

rule foo = 
{
    print bar

    print baz(0)
    print baz(-1)
    print baz(1)
    
    print qux(5, 12)

    call ( prog_mem( 2 ) )

    let x = 3 in
        call ( prog_mem( x ) )
    
    program( self ) := undef
}

rule bbb = skip

rule ccc = skip