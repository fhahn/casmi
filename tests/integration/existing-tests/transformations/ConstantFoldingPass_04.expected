foo=STATEMENTS(LET CONST(ident c0__,num 1234),STATEMENTS(LET CONST(ident c1__,num 3210),STATEMENTS(LET CONST(ident c2__,num -1),STATEMENTS(LET CONST(ident c3__,num 4321),STATEMENTS(LET CONST(ident c4__,num 1987),STATEMENTS(LET CONST(ident c5__,num 3),PARBLOCK(STATEMENTS(PRINT(LIST(ident c0__,null),null),STATEMENTS(PRINT(LIST(ident c1__,null),null),STATEMENTS(PRINT(LIST(LOOKUP(ident baz,LIST(ident c2__,null)),null),null),STATEMENTS(PRINT(LIST(ident c3__,null),null),STATEMENTS(PRINT(LIST(ident c4__,null),null),STATEMENTS(CALL_INDIRECT(ident bbb,null),STATEMENTS(PARBLOCK(CALL_INDIRECT(ident ccc,null),null),STATEMENTS(UPDATE(LOOKUP(ident program,LIST(SELF,null)),UNDEF),null)))))))),null)))))))

bbb=SKIP

ccc=SKIP